Management of membergroup icons

Akyhne opened the pull request:
http://www.simplemachines.org/community/index.php?topic=490266.msg3438509#msg3438509

This pull adds an extra sub menu to the "Manage Membergroups".

The page is used to edit and manage membergroup and post based group's icons.

Reason:
At the moment, it is only possible to add new icons via FTP. It is only possible to change a MG or PB group's icon, by knowing it's exact name.
It is not possible to see if the icon is present in all themes in the forum.
In the edit page of a group, where you change the icon, you could use the $language to apparently change the path to the uer's language. This didn't work at all.

New with this modification:
On the page where you changed the icon, the textbox has been replaced by a hidden input and an image showing the current selected image.
Clicking the image, will call a dropdown that works similar to the way you change a posts icon (xx.png etc.). It uses jQuery.
The "option" to use $language as a part of the file path, has been removed. It had no effect.

The new page added, gives you the option to copy icons between different themes with drag'n drop (jQuery). You delete an image, by doubleclicking it.
New images can be uploaded and you can decide whether the image should be copied to any theme or to a specific theme. So you can still have different icons with the same name for different themes.
You can rename the file during upload (no conversion though) and you can choose whether existing files should be replaced.

To make all this work, the SMF forum software needs to have a separate folder for MG & PB group icons. And why not?! Almost all images have their own subfloder in the theme.
I chose to rename the subfolder to /default/images/membergroup_icons.
These images needs to be moved there in the default theme:
blank.png
icon.png
iconadmin.png
icongmod.png
iconmod.png

This commit doesn't contain these changes, cause I can' figure out how to upload folders and delete files.

Known Issues:
For good reasons, you can't use the drag and doubleclick features on smartphones.
The only thing that needs to be taken care of, besides moving the icons, is to look on how the upload handles errors and what steps should be taken if the folder membergroup_icons is not writable.

Akyhne committed:
membergroup icons

Akyhne committed:
membergroup icons

Akyhne committed:
membergroup icons

Akyhne committed:
membergroup icons

Akyhne committed:
membergroup icons

Akyhne committed:
membergroup icons

Akyhne committed:
membergroup icons

Akyhne committed:
membergroup icons

Akyhne committed:
Membergroup icons

Oops, copy & paste issue. There are no conditions, as it needs to fetch all themes in the DB.

norv commented on Sources/ManageMembergroups.php:
```diff
@@ -1180,4 +1194,172 @@ function ModifyMembergroupsettings()
+	// Copy an image
+	if (isset($_REQUEST['copy']))
+	{
+
```
checkSession();

Akyhne replied:
I see it (now)!

norv commented on Sources/ManageMembergroups.php:
```diff
@@ -1180,4 +1194,172 @@ function ModifyMembergroupsettings()
+	// Delete doubleclicked image
+	if (isset($_REQUEST['delete']))
+	{
+		$imgfile = $context['available_themes'][$_POST['delete_theme']]['theme_dir']. '/images/membergroup_icons/'. $_POST['delete_img'];
```
checkSession();

norv commented on Sources/ManageMembergroups.php:
```diff
@@ -1180,4 +1194,172 @@ function ModifyMembergroupsettings()
+
+	// Upload new file(s)
+	if (isset($_REQUEST['save']))
+	{
```
missing a checkSession();

Akyhne replied:
Fixed!

norv commented on Sources/ManageMembergroups.php:
```diff
@@ -1180,4 +1194,172 @@ function ModifyMembergroupsettings()
+		$y++;
+	}
+	$context['post_url'] = $scripturl . '?action=admin;area=membergroups;save;sa=icons';
+	createToken('admin-mmg');
```
this token should be then validated when the form is submitted

Akyhne replied:
I tried to get those tokens to work, but they won't validate and gives me the error: Your session timed out while posting. Please go back and try again.

I have no idea how they work (except it's a sequrity fix) and how to fix it.

Akyhne replied:
Ah, I just realised I have to put two hidden fields in the bottom of every form that can be submitted. Then no errors!

norv replied:
Yay! Thanks :)

norv commented on Sources/ManageMembergroups.php:
```diff
@@ -1180,4 +1194,172 @@ function ModifyMembergroupsettings()
 	prepareDBSettingContext($config_vars);
 }
 
+function ModifyMembergroupicons()
```
ModifyMembergroupIcons()
[/pedantry] :)

Akyhne replied:
Sir, are you just bored or what? lol

Will fix!

norv replied:
On keeping naming patterns consistent in SMF? Always! (ask @Spuds :laughing: )

Akyhne committed:
Membergroup Icons - managemembergroups.php

Added checksession()

norv commented on Sources/ManageMembergroups.php:
```diff
@@ -1180,4 +1194,172 @@ function ModifyMembergroupsettings()
+		if ($_POST['delete_img'] != "" && (int)$_POST['delete_theme'] > 0)
+		{
+			echo ' Path: \'', $context['available_themes'][(int)$_POST['delete_theme']]['theme_dir'], '/images/membergroup_icons/', $_POST['delete_img'], '\'';
+			if (file_exists($imgfile))
```
file_exists() is not necessary, @unlink($imgfile) on the next line should be enough

Akyhne replied:
I just added it, because I don't know how many kind of errors you want to check for.

Akyhne committed:
Membergroup Icons - M-Membergroups.template.php

Added token session check.
Fixed release of dragged icon should always fall back in position. Fixed using jquery-ui 1.9.1 as 1.7.1 had issues with several browsers (official bug)

norv commented on Sources/ManageMembergroups.php:
```diff
@@ -1124,6 +1136,20 @@ function EditMembergroup()
 		$context['inheritable_groups'][$row['id_group']] = $row['group_name'];
```
Debugging leftover? :)

Akyhne replied:
What part?!

norv replied:
Hah! You are too fast for me lol, lines changed

``` php
echo ' Path: \'', $context['available_themes'][(int)$_POST['delete_theme']]['theme_dir'], '/images/membergroup_icons/', $_POST['delete_img'], '\'';
```

Akyhne replied:
Fixed!

norv commented on Sources/ManageMembergroups.php:
```diff
@@ -1180,4 +1194,172 @@ function ModifyMembergroupsettings()
+		if (!in_array(strtolower(substr(strrchr($_POST['copy_image'], '.'), 1)), $allowedTypes))
+			fatal_lang_error('smileys_upload_error_types', false, array(implode(', ', $allowedTypes)));
+
+		$img_source = $context['available_themes'][$_POST['copy_from_theme']]['theme_dir']. '/images/membergroup_icons/'. $_POST['copy_image'];
```
(int) $_POST['copy_from_theme']

Akyhne committed:
Manage membergroup icons

Fixed "protection" of integers

Akyhne committed:
Memberg. icons - Managemebbergroups.php

Oops, an echo had occured! Added cola and cookies.

Akyhne committed:
MG icons - admin.css

It's better if the dropdown shows icons in multiple rows, if more than 6-7 icons are present.

Akyhne committed:
I didn't do that (did I?)

Akyhne committed:
...

Akyhne committed:
Line break before the end

Akyhne committed:
MG Icons 

Small layout issue

Akyhne committed:
MG Ions

Man! You shouldn't copy/paste to this site! So many extra line breaks inserted GGGRRRRRrrrrr!!!!!

emanuele45 commented:
Nitpicking bureaucracy: none of the commits is signed-off as described in the [how to contribute](https://github.com/SimpleMachines/SMF2.1#how-to-contribute) section of the readme.

Spuds commented:
And its on the development branch .... don't know if that was intended or not

Akyhne commented:
That was intended.

Look! You can use the code, the idea or drop it if you want. I did my part.

Oldiesmann commented:
If you want to get this in, resubmit it to the release-2.1 branch with all of the commits signed off (git commit -asm 'commit message'). I'd hate to see you not get credit for your work :)

Akyhne commented:
Oldiesmann. I deleted my account a while back on simplemachines.org
I just lost interest in the SMF project and don't know how to apply this code to the 2.1 branch.
Feel free to use it if you like. I don't need to be credited.

Oldiesmann commented:
If I get some time I'll take a look at it.

Akyhne commented:
Thanks :-)

Oldiesmann commented:
Looking into this now, and have a few questions before I get to adding it in...
1. How does this handle existing groups' icons?
2. Can it still be customized for each user's language? I know this isn't a widely used feature, but I'd like to avoid breaking existing features if possible.
3. Does this handle errors yet, or does that still need to be coded?
4. Does it fully work if JS is disabled?

Akyhne commented:
1) It just enhances the way it Work now, by adding the ability to upload, delete and change icons.
2) No
3) It has been prepared to a certain level, but not finished as I thought it should be up to the devs to choose the level of errors
4) No

Oldiesmann commented:
So they can't manage membergroup icons at all without JS enabled? That might be an issue unfortunately. I realize most people do have JavaScript enabled, but I'm not sure I want to add such a useful feature if it won't work without JavaScript.

ghost commented:
is there another way to do it without javascript?

Oldiesmann commented:
Any advice on this @Akyhne? I'd rather not remove the ability to manage membergroup icons without JS

Akyhne commented:
Well, it should be possible to keep the original way of changing membergroup icons, if no Javascript is detected.

Akyhne commented:
If you want to see how this works, check http://www.youtube.com/watch?v=AKI5RZB_JFw

MissAllSunday commented:
Is this still on the plans for 2.1?  It can be done and still offer support for users with no JS using unobtrusive JS but 1) It would have to be redone and 2) There some issues we will have to address if we really want to implement this, I remember posted a few ones in the bug report about this on mantis.

Oldiesmann commented:
I think it's a nice addition, so if it can be done without absolutely requiring JS, then we should definitely try to get it in.

matthew-kerle commented:
You can almost _always_ write things in such a manner (especially utilizing event.preventDefault();) to handle things if the JavaScript doesn't. Definitely possible here.

Arantor commented:
I don't really understand the reason to discourage use of JS in the admin panel. Front side, sure, but the admin panel, not so much.

Akyhne commented:
I agree with Arantor. As admin, you shouldn't have to protect yourself against your own forum software.
But if it's for functionality only, it should be fairly easy to implement both.
The way I made the system, you can still upload files manually as in the old days, you just can't add/edit/delete icons. But if the current textbox is kept to enter the name o the file manually when JS is not present, you still keep the old functionanlity.

MissAllSunday commented:
I suppose we can use this and wait for people's reactions to it, Beta release would be perfect for that.

The are still some issues with using this PR though,
- Although correct, this was based on development branch, it would be difficult to merge it to release branch.
- Non of the commits are signed up.

MissAllSunday commented:
Another issue with using this is that Akyhne wrote directly into the development branch rather than creating a new one, I tried to cherry pick all of his commits one by one into a new branch based on release but since this was based directly on the development branch, git is trying to merge the entire development branch rather than just the single commit.

emanuele45 commented:
rebase is your friend.

MissAllSunday commented:
> rebase is your friend.
> 
> —

Care to elaborate? And please do not just copy/paste some url, I'm already  
aware of rebase but their explanations aren't that clear for me.

Akyhne commented:
I did as described on simplemachines.org - created a copy of the dev. branch, made my changes and commited them to the SMF dev branch.
Some say I did it the wrong way, but I'm not familiar with Github at all.

XinYenFon commented:
You can get a PR as patch and apply it via terminal to your local repo.

https://github.com/SimpleMachines/SMF2.1/pull/230.patch

MissAllSunday commented:
Thanks Antes, I tried that but there were a lot of errors and the patch did not apply.  As for rebasing, I haven't found any meaningful example on how to do it, I have no idea if the branch I need to rebase is my current branch or the development one.

I've been looking at some docs from Drupal, it seems they only handle contributions via patches so they have some documentation about it, unfortunately, it is heavily focused on their own issues and common problems.

Oh well, someone more versed on rebasing can do this way better than I am.

emanuele45 commented:
Jessica González wrote:

> > rebase is your friend.
> 
> Care to elaborate? And please do not just copy/paste some url, I'm
> already
> aware of rebase but their explanations aren't that clear for me.

It's not rocket science: git rebase and fix the conflicts.
There is nothing else to do:
https://github.com/emanuele45/playpen/tree/akyhne_devel
then remove the commits that are not supposed to be there (Antes and
Norv) and since anyway the commit messages were kind of useless just
squash everything into one big commit and:
git commit -s --amend

Obviously not tested (not even for php syntax).

MissAllSunday commented:
> Jessica González wrote:
> 
> > > rebase is your friend.
> > 
> > Care to elaborate? And please do not just copy/paste some url, I'm
> > already
> > aware of rebase but their explanations aren't that clear for me.
> 
> It's not rocket science: git rebase and fix the conflicts.
> There is nothing else to do:
> https://github.com/emanuele45/playpen/tree/akyhne_devel
> then remove the commits that are not supposed to be there (Antes and
> Norv) and since anyway the commit messages were kind of useless just
> squash everything into one big commit and:
> git commit -s --amend
> 
> Obviously not tested (not even for php syntax).
> —

Please save to yourself your snarky/aggressive comments... its getting  
annoying that you always reply to me like that, if you don't like me for  
whatever reason then simply ignore me.

I asked in good faith, if it wasn't difficult for me I would have done it  
already... anyway, if it is soo easy for you then do the honors and finish  
this up, I'm off.
